File: app/src/services/usda-food.ts
import axios from "axios";
import { usda_api } from "src/constants/routes/usda-api";
import { Food } from "src/types/food";
import { FoodDto, SearchResultDto } from "src/types/usda-food";
import { usdaFoodToFood, usdaSearchFoodToFood } from "src/utils/usda-food";

const apiKey = 'g8hkV8fS6A17dTaS0DEk464LisJCu8AdN2gKIU2C';

export const getFood = async (searchTerm: string, pageNumber: number) => {
    const res = await axios
        .get<SearchResultDto>(usda_api.search, {
            params: {
                api_key: apiKey,
                generalSearchInput: searchTerm,
                dataType: 'Branded',
                pageNumber: pageNumber,
                pageSize: 10,
            },
        });

    const data = res.data;

    return data.foods.map((food) => usdaSearchFoodToFood(food));
}

export const getFoodById = async (id: string) => {
    const res = await axios
        .get<FoodDto>(usda_api.food + '/' + id, {
            params: {
                api_key: apiKey,
            },
        });

    const data = res.data;

    return usdaFoodToFood(data);
}

export const getFoodByIds = async (ids: number[]) => {
    const res = await axios
        .get<FoodDto[]>(usda_api.foods, {
            params: {
                api_key: apiKey,
                fdcIds: ids.join(','),
            },
        });

    const data = res.data;

    return data.map((food) => usdaFoodToFood(food));
}

File: backend-v2/src/services/usda_food.rs
use serde::Deserialize;
use surf::Error;

use crate::dto::food_dtos::FoodDto;

static USDA_DOMAIN: &str = "https://api.nal.usda.gov/fdc/v1";
static USDA_API_KEY: &str = "g8hkV8fS6A17dTaS0DEk464LisJCu8AdN2gKIU2C";

enum Routes {
    Search,
    FoodDto,
    Foods,
}

impl ToString for Routes {
    fn to_string(&self) -> String {
        match self {
            Routes::Search => format!("{USDA_DOMAIN}/foods/search"),
            Routes::FoodDto => format!("{USDA_DOMAIN}/food"),
            Routes::Foods => format!("{USDA_DOMAIN}/foods"),
        }
    }
}

#[derive(Debug, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct USDANutrientDto {
    pub id: Option<i32>,
    pub number: Option<String>,
    pub name: Option<String>,
    pub unit_name: Option<String>,
}

#[derive(Debug, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct USDAFoodNutrientDto {
    pub id: Option<i32>,
    pub amount: Option<f32>,
    pub nutrient: USDANutrientDto,
}

#[derive(Debug, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct USDABranndedFoodItemDto {
    pub fdc_id: i32,
    pub description: String,
    pub data_type: String,
    pub brand_name: Option<String>,
    pub brand_owner: Option<String>,
    pub household_serving_full_text: Option<String>,
    pub ingredients: Option<String>,
    pub serving_size: Option<f32>,
    pub serving_size_unit: Option<String>,
    pub branded_food_category: Option<String>,
    pub food_nutrients: Vec<USDAFoodNutrientDto>,
}

pub async fn get_usda_food_by_search(search: &str, page: i32) -> Result<(), Error> {
    #[derive(Serialize)]
    #[serde(rename_all = "camelCase")]
    struct SearchQuery {
        #[serde(rename = "api_key")]
        api_key: String,
        general_search_input: String,
        data_type: String,
        page_number: i32,
        page_size: i32,
    }

    let res = surf::get(Routes::Search.to_string())
        .query(&SearchQuery {
            api_key: USDA_API_KEY.to_string(),
            general_search_input: search.to_string(),
            data_type: "Branded".to_string(),
            page_number: page,
            page_size: 10,
        })?
        .recv_json::<Vec<USDABranndedFoodItemDto>>()
        .await?;

    Ok(())
}

pub async fn get_usda_food_by_id(id: &str) -> Result<FoodDto, Error> {
    #[derive(Serialize)]
    struct FoodQuery {
        api_key: String,
    }

    let res = surf::get(format!("{}/{}", Routes::FoodDto.to_string(), id))
        .query(&FoodQuery {
            api_key: USDA_API_KEY.to_string(),
        })?
        .recv_json::<USDABranndedFoodItemDto>()
        .await?;

    let res = FoodDto::from(res);

    Ok(res)
}

pub async fn get_usda_foods_by_ids(ids: Vec<i32>) -> Result<Vec<FoodDto>, Error> {
    #[derive(Serialize)]
    #[serde(rename_all = "camelCase")]
    struct FoodsQuery {
        #[serde(rename = "api_key")]
        api_key: String,
        fdc_ids: String,
    }

    let ids = &ids
        .iter()
        .map(|id| id.to_string())
        .collect::<Vec<String>>()
        .join(",");

    let res = surf::get(Routes::Foods.to_string())
        .query(&FoodsQuery {
            api_key: USDA_API_KEY.to_string(),
            fdc_ids: ids.to_string(),
        })?
        .recv_json::<Vec<USDABranndedFoodItemDto>>()
        .await?;

    let res = res.into_iter().map(FoodDto::from).collect::<Vec<_>>();

    Ok(res)
}
